[Core] Enhance `Response` and `RequestContent` for Low Level Client

ellismg opened the pull request:
- Add a `Content` property to `Response` which returns a `BinaryData`
  that wraps the content of the response.

- Add factory methods to `RequestContent` to create instances from
  `BinaryData`, an object (using an `ObjectSerializer`), or a `string`
  (in the string case we convert the data to UTF-8).

ellismg committed:
[Core] Enhance `Response` and `RequestContent` for Low Level Client

- Add a `Content` property to `Response` which returns a `BinaryData`
  that wraps the content of the response. This property only returns a
  value when the content stream is null (in which case we return an
  empty `BinaryData`) or when the content stream is backed by a memory
  stream and we can extract the memory stream's buffer. It happens
  that this is the common case. In other cases we throw, as we are
  unwilling to do IO in this property.

- Add factory methods to `RequestContent` to create instances from
  `BinaryData`, an object (using an `ObjectSerializer`), or a `string`
  (in the string case we convert the data to UTF-8).

ellismg commented:
(Opening as a draft for now)

Here's some updates to `Response` and `RequestContent` that we have been discussing, wanted to use the PR as a way for us to discuss some stuff.  See comments.

ellismg commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,42 @@ public abstract class Response : IDisposable
+        /// <summary>
+        /// Gets the contents of HTTP response. Returns a BinaryData with no payload for responses without content.
+        /// </summary>
+        public virtual BinaryData Content
```
Are there any concerns here about the case where this method ends up doing IO in the case where the content stream has not been already read into a `MemoryStream`?

pakrym replied:
Yes, I think this property should be assigned in the https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs and throw in the getter when not assigned.

KrzysztofCwalina replied:
I am reading the thread and it seems like we agreed to return empty BD instead of throwing.

Separately, don't like the policy doing two allocations here (BD and Stream). Can policy allocate just one; the other one would be allocated by concrete response? This would allow concrete responses to optimize their storage.

pakrym replied:
Making it nullable would make everyone check it for null if they have nullability enabled. And if we are talking about the same thread it still seems to me we agreed on exception:

You said:
>OK. If we discover that users see this as very problematic, we can always expose Response.IsStreaming property.


>Separately, don't like the policy doing two allocations here (BD and Stream). Can policy allocate just one; the other one would be allocated by concrete response? This would allow concrete responses to optimize their storage.

Yeah, I'm not really a fan of hardcoding a special case for memorystream in Response but maybe it's fine.

KrzysztofCwalina replied:
There was another thread where Matt suggested returning empty and Johan agreed that null vs empty content is not an important distinction.

>Making it nullable would make everyone check it for null if they have nullability enabled. 

Nobody wants to make it nullable.

KrzysztofCwalina replied:
If we were designing the behavior today, I would return empty stream and if we discover somebody needs to know the difference between null and empty, we would add a property to check that.

pakrym replied:
I don't understand. Returning an empty stream when the response is not buffered? How does that help?

KrzysztofCwalina replied:
Sorry, this thread went from null vs empty stream to streaming responses. I was just referring to the "no content" vs empty content distinction.

ellismg commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,42 @@ public abstract class Response : IDisposable
+                if (memoryContent == null)
+                {
+                    memoryContent = new MemoryStream();
+                    ContentStream.CopyTo(memoryContent);
```
Should we be trying to do anything more here?  If the user already read a few bytes from the Stream, we wouldn't end up copying everything, but we couldn't really seek to the start of the stream in all cases (we could do this in the case where CanSeek returned `true`, I guess).

ellismg commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,42 @@ public abstract class Response : IDisposable
+            {
+                if (_content != null)
+                {
+                    return _content;
```
The `ContentStream` property of `Response` can be `set` (but I'm not really sure in what cases that ends up happening).  For our implementations, would we want to throw away this cached value in the case where the stream is replaced (so if after replacing the stream you call `.Content` you see the new content)?

KrzysztofCwalina replied:
I think we have no choice but to throw it away. But then how do we read the stream in a sync context (if it's not a memory stream). I guess we just sync over async. This is a bit of a mess.

ellismg commented on sdk/core/Azure.Core/src/RequestContent.cs:
```diff
@@ -52,6 +57,41 @@ public abstract class RequestContent : IDisposable
+        /// <param name="content">The <see cref="string"/> to use.</param>
+        /// <returns>An instance of <see cref="RequestContent"/> that wraps a <see cref="string"/>.</returns>
+        /// <remarks>The returned content represents the UTF-8 Encoding of the given string.</remarks>
+        public static RequestContent Create(string content) => Create(UTF8NoBomEncoding.GetBytes(content));
```
@KrzysztofCwalina I suspect you intended that this would do a conversion to UTF-8 since that's what we wanted to send on the wire, but just checking.  Do you think it would be worthwhile to have an optional Encoding parameter and if it's null assume UTF-8 (for cases where say you want to send a string in UTF-16 or some other encoding as part of an "upload" operation or something?)

ellismg commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,42 @@ public abstract class Response : IDisposable
+
+                if (memoryContent == null)
+                {
+                    memoryContent = new MemoryStream();
```
Do folks think it is worthwhile to inspect the `Content-Length` header here to try to pre-size the buffer, or in cases where we have a content length will the entire response have already been read into a MemoryStream?

pakrym replied:
>, or in cases where we have a content length will the entire response have already been read into a MemoryStream?

No, storage download responses have content length but are streamed. I think you should only create binary data if the stream is already a memory stream.

ellismg replied:
> I think you should only create binary data if the stream is already a memory stream.

Just to confirm my understanding of the proposal: This would mean that we would make the `Content` property nullable and return `null` in these cases?

pakrym replied:
The conversation I took part in decided to throw from Content for non-buffered responses. It seems there was another one that decided to return empty BD instead. What we definitely shouldn't do is synchronously read from live network streams.

pakrym commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,42 @@ public abstract class Response : IDisposable
         /// </summary>
         public virtual ResponseHeaders Headers => new ResponseHeaders(this);
 
+        private BinaryData? _content;
```
Why would we not want to cache the instance? I don't think we want to allocate every time somebody pulls onto the property.

KrzysztofCwalina replied:
We do want to cache it in our concrete implementations, not in the abstraction. The abstraction should allow implementations that do their own memory management, e.g. use pooled memory for buffers.

pakrym replied:
We already do memory management in ResponseBodyPolicy (allocating memory stream and buffering).

pakrym replied:
I don't think having a base overrideable implementation is making it worse.

KrzysztofCwalina replied:
You cannot override fields.

pakrym replied:
You can override the Content property.

ellismg replied:
We could make the field `protected` so subclasses could at least reuse the underlying field (I suspect that our subclasses will want to null the cached value out if the `ContentStream` setter).  Does that seem reasonable or should we just accept that most subclasses that would want to customize the behavior here will end up ignoring the `_content` field and the underlying Response object will be a pointer size larger than it needs to be?

pakrym replied:
>I suspect that our subclasses will want to null the cached value out if the ContentStream setter

Hm, yeah. This kind of makes the base caching implementation incorrect in the general case. Because it still requires every implementor to opt into invalidating the cache.

Taking that into account I think we should return a new instance of the BinaryData in the base class and not cache.

ellismg replied:
> Taking that into account I think we should return a new instance of the BinaryData in the base class and not cache. That would at least make the implementation correct and we can make it efficient in the individual implementations.

SGTM!

KrzysztofCwalina commented on sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs:
```diff
@@ -447,11 +449,40 @@ public PipelineResponse(string requestId, HttpResponseMessage responseMessage, S
+
+                    if (ContentStream == null)
+                    {
+                        _content = new BinaryData(Array.Empty<byte>());
```
It would be good if we added BinaryData.Empty<T>

ellismg replied:
Will do.

KrzysztofCwalina commented on sdk/core/Azure.Core/src/Pipeline/HttpClientTransport.cs:
```diff
@@ -447,11 +449,40 @@ public PipelineResponse(string requestId, HttpResponseMessage responseMessage, S
+                    if (contentMemoryStream == null)
+                    {
+                        contentMemoryStream = new MemoryStream();
+                        ContentStream.CopyTo(contentMemoryStream);
```
I don't think we want to do it for very large and non seekable streams, e.g. the network stream. I think we should either throw for non-buffered streams, or check if the stream is seekable and has small length first.

KrzysztofCwalina replied:
Also, I wonder if we should not change the buffered policy to work as follows: it buffers to BinaryData and stores it in response. The stream property is lazy and returns a stream wrapping the BD. The main reason for this is that is more optimal for our JsonDocument.Parse method (i.e. more efficient to parse Memory<byte>) [I think]

ellismg commented:
Pushed some updates here based on our conversation earlier today.  There we agreed that `Content` would be non-nullable, but we would throw `InvalidOperationException` in cases where the ContentStream is not a `MemoryStream`.  In addition, I discovered that in some cases calling `GetBuffer()` on `MemoryStream` would fail (depending on how the stream is constructed) so we also now throw `InvalidOperationException` in these cases as well (fortunately, our usage of MemoryStream in the pipeline yields a stream we can always take the buffer from).

Because we no longer read the stream, I think it is sufficient for us to just put the code in the abstraction and not worry about caching in the implementations.  We would end up bloating all the responses by a pointer size and we don't really gain much.  In practice, I suspect people will call `.Content` at most once and if they do call it multiple times we end up doing minimal extra work (but we do have to allocate a new BinaryData each time).

KrzysztofCwalina commented on sdk/core/Azure.Core/src/RequestContent.cs:
```diff
@@ -15,6 +17,9 @@ namespace Azure.Core
     public abstract class RequestContent : IDisposable
     {
+        private static readonly Encoding s_UTF8NoBomEncoding = new UTF8Encoding(false);
+        private static readonly ObjectSerializer s_DefaultSerializer = new JsonObjectSerializer();
```
We should add a static like that to JsonObjectSerializer. Otherwise we will end up with many of such statics all over Azure.Core.

ellismg replied:
Done.

KrzysztofCwalina commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,39 @@ public abstract class Response : IDisposable
+
+                if (memoryContent == null)
+                {
+                    throw new InvalidOperationException($"The {nameof(ContentStream)} is not a ${typeof(MemoryStream)}");
```
I think we should change the message. The user won't know why it's not memory stream. Maybe we should just say that this response contains non-seekable stream.

ellismg replied:
I will make the change, but I think no matter what if the user hits this exception they are going to be confused as to why it didn't work.  It's not any clearer as to why the stream isn't seakable vs why it isn't a memory stream and the stream being seekable isn't actually sufficient for this property to return a BinaryData.  The stream *must* be of type `MemoryStream` and we *must* be able to access it's buffer.

pakrym commented on sdk/core/Azure.Core.TestFramework/src/MockResponse.cs:
```diff
@@ -31,7 +31,7 @@ public MockResponse(int status, string reasonPhrase = null)
         public void SetContent(byte[] content)
         {
-            ContentStream = new MemoryStream(content);
+            ContentStream = new MemoryStream(content, 0, content.Length, true, true);
```
Do we need it writable?

ellismg replied:
We don't need it writable, I can change that.  We do need it to be publicly exposable.

pakrym commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,39 @@ public abstract class Response : IDisposable
+            {
+                if (ContentStream == null)
+                {
+                    return new BinaryData(Array.Empty<byte>());
```
WE might want to cache this binarydata instance. Maybe add BinaryData.Empty.

ellismg replied:
I will add a `BinaryData.Empty` property and use it.

pakrym commented on sdk/core/Azure.Core/tests/ResponseTests.cs:
```diff
@@ -87,6 +90,32 @@ public void ToStringsFormatsStatusAndMessageForNoBodyResponse()
+        }
+
+        [Test]
+        public void ContentPropertyThrowsForStream()
```
```suggestion
        public void ContentPropertyThrowsForNonMemoryStream()
```

pakrym commented on sdk/core/Azure.Core/tests/ResponseTests.cs:
```diff
@@ -87,6 +90,32 @@ public void ToStringsFormatsStatusAndMessageForNoBodyResponse()
+            responseWithBody.SetContent("body content");
+            Assert.AreEqual("body content", responseWithBody.Content.ToString());
+
+            // Ensure that the BinaryData is formed over the used portion of the memory stream, not the entire buffer.
```
SHouldn't it be the opposite?

ellismg replied:
Maybe I wrote the comment in a bad way.  My goal in this test is that we ensure the `BinaryData` that is retuned from `.Content` only contains data from the valid part of the stream (and not the view of the entire underlying buffer).  In this test we create a MemoryStream with a backing buffer of 50 bytes, don't write anything to it and then confirm the Content returned is empty (and not 50 bytes of 0's).

pakrym replied:
I'm worried about a different scenario:

We get a response, the client reads the content stream to deserialize it (https://github.com/Azure/azure-sdk-for-net/issues/17711) and doesn't rewind it back. What would the `Content` be in that case?

ellismg replied:
> What would the Content be in that case?

Ahh - I understand.  I think that in all these cases we want `Content` to be the entire contents of the stream, correct?  Having it mean: "read from the current stream position to the end of the stream" feels wrong?

pakrym replied:
>I think that in all these cases we want Content to be the entire contents of the stream, correct


Yep, I think so too.

ellismg replied:
Added another test here to ensure we have the right behavior

pakrym commented on sdk/core/Azure.Core/tests/ResponseTests.cs:
```diff
@@ -87,6 +90,32 @@ public void ToStringsFormatsStatusAndMessageForNoBodyResponse()
+            response.ContentStream = new ThrowingStream();
+
+            Assert.Throws<InvalidOperationException>(() => { BinaryData d = response.Content; });
+        }
```
+ Test for non-public MemoryStream buffer.

pakrym commented on sdk/core/Azure.Core/tests/TransportFunctionalTests.cs:
```diff
@@ -38,6 +38,8 @@ public TransportFunctionalTests(bool isAsync) : base(isAsync)
+                new object[] { RequestContent.Create(new BinaryData(new byte[10])), 10 },
             };
 
         [TestCaseSource(nameof(ContentWithLength))]
```
Let's add a few Content tests to the https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs to make sure things work E2e

ellismg replied:
Added some more asserts to existing tests around `ContentStream` and wrote a new one to ensure the correct thing happens  with `ExtractResponseContent()`

pakrym commented on sdk/core/Azure.Core/api/Azure.Core.netstandard2.0.cs:
```diff
@@ -314,12 +315,17 @@ public abstract partial class Request : System.IDisposable
     public abstract partial class RequestContent : System.IDisposable
     {
         protected RequestContent() { }
+        public static Azure.Core.RequestContent Create(System.BinaryData content) { throw null; }
```
`+` Tests for new method here.

https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/tests/HttpPipelineRequestContentTests.cs

pakrym commented on sdk/core/Azure.Core/src/Response.cs:
```diff
@@ -41,6 +41,39 @@ public abstract class Response : IDisposable
+        /// <remarks>
+        /// Throws <see cref="InvalidOperationException"/> when <see cref="ContentStream"/> is not a <see cref="MemoryStream"/>.
+        /// </remarks>
+        public virtual BinaryData Content
```
Do we want to override implementations for existing transports as part of this PR as well?

ellismg replied:
I think the only reason to override it would be if we wanted to cache the binary data, so we didn't have to allocate a new BinaryData object on every call to `.Content`.

I am wondering if in practice that's the right choice to make.  It would cause another field on the `Response` type, to save an allocation in the case where you call `.Content` more than once on a `Response`.  I'm not sure that's worthwhile especially considering that when using the higher level abstractions like `Response<T>` you often don't even touch the response.

If you think it is worthwhile, I am happy to do it (or if there's something better we could do in the concrete implementations that we can't do in the abstraction, let me know) but that is why I did not override implementations for the existing transports.

pakrym replied:
Let's start by not caching.

pakrym commented on sdk/core/Azure.Core/src/Serialization/JsonObjectSerializer.cs:
```diff
@@ -20,6 +20,11 @@ public class JsonObjectSerializer : ObjectSerializer, IMemberNameConverter
+        /// <summary>
+        /// A shared instance of <see cref="JsonObjectSerializer"/>, initialized with the default options.
+        /// </summary>
+        public static readonly JsonObjectSerializer Default = new JsonObjectSerializer();
```
I think this should be a property.

ellismg replied:
Done!

pakrym commented on sdk/core/Azure.Core/tests/HttpPipelineRequestContentTests.cs:
```diff
@@ -48,5 +46,61 @@ public void StreamCopyTo()
+
+            for (int i = 0; i < size; i++)
+            {
+                Assert.AreEqual(sourceArray[i], destinationArray[i]);
```
CollectionAssert.AreEqual

pakrym commented on sdk/core/System.Memory.Data/src/BinaryData.cs:
```diff
@@ -16,6 +16,11 @@ namespace System
+        /// <summary>
+        /// A <see cref="BinaryData"/> instance for an empty byte array.
+        /// </summary>
+        public static readonly BinaryData Empty = new BinaryData(Array.Empty<byte>());
```
Should be a property too.

ellismg committed:
Update TextAnalytics Low Level Client to use new patterns

ellismg commented:
@pakrym Marked this as ready for review, I've addressed all your feedback from the previous rounds.  I also have included updates to the TA hand authored prototype to use these.

pakrym commented:
Can you add changelog entries for the new APIs and re-run the Export-APi script.
